registry/kubernetes/registry.go

```diff
@@ -0,0 +1,369 @@
+	"strings"
+	"sync"
+	"time"
+)
+
+import (
+	gxnet "github.com/dubbogo/gost/net"
+	perrors "github.com/pkg/errors"
+)
+
+import (
+	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/common/constant"
+	"github.com/apache/dubbo-go/common/extension"
+	"github.com/apache/dubbo-go/common/logger"
+	"github.com/apache/dubbo-go/registry"
+	"github.com/apache/dubbo-go/remoting/kubernetes"
+)
+
+var (
+	processID = ""
+	localIP   = ""
+)
+
+const (
+	Name              = "kubernetes"
+	RegistryConnDelay = 3
+)
+
+func init() {
+	processID = fmt.Sprintf("%d", os.Getpid())
+	localIP, _ = gxnet.GetLocalIP()
+	extension.SetRegistry(Name, newKubernetesRegistry)
+}
+
+type kubernetesRegistry struct {
+	*common.URL
+	birth int64 // time of file birth, seconds since Epoch; 0 if unknown
+
+	cltLock  sync.Mutex
+	client   *kubernetes.Client
+	services map[string]common.URL // service name + protocol -> service config
+
+	listenerLock   sync.Mutex
+	listener       *kubernetes.EventListener
+	dataListener   *dataListener
+	configListener *configurationListener
+
+	wg   sync.WaitGroup // wg+done for kubernetes client restart
+	done chan struct{}
+}
+
+func (r *kubernetesRegistry) Client() *kubernetes.Client {
+	return r.client
+}
+func (r *kubernetesRegistry) SetClient(client *kubernetes.Client) {
+	r.client = client
+}
+func (r *kubernetesRegistry) ClientLock() *sync.Mutex {
+	return &r.cltLock
+}
+func (r *kubernetesRegistry) WaitGroup() *sync.WaitGroup {
+	return &r.wg
+}
+func (r *kubernetesRegistry) GetDone() chan struct{} {
+	return r.done
+}
+func (r *kubernetesRegistry) RestartCallBack() bool {
+
+	services := []common.URL{}
+	for _, confIf := range r.services {
+		services = append(services, confIf)
+	}
+
+	for _, confIf := range services {
+		err := r.Register(confIf)
+		if err != nil {
+			logger.Errorf("(kubernetesProviderRegistry)register(conf{%#v}) = error{%#v}",
+				confIf, perrors.WithStack(err))
+			return false
+		}
+		logger.Infof("success to re-register service :%v", confIf.Key())
+	}
+	return true
+}
+
+func newKubernetesRegistry(url *common.URL) (registry.Registry, error) {
+
+	r := &kubernetesRegistry{
+		URL:      url,
+		birth:    time.Now().UnixNano(),
+		done:     make(chan struct{}),
+		services: make(map[string]common.URL),
+	}
+
+	if err := kubernetes.ValidateClient(r); err != nil {
+		return nil, err
+	}
+
+	r.wg.Add(1)
+	go kubernetes.HandleClientRestart(r)
+
+	r.listener = kubernetes.NewEventListener(r.client)
+	r.configListener = NewConfigurationListener(r)
+	r.dataListener = NewRegistryDataListener(r.configListener)
+
+	return r, nil
+}
+
+func (r *kubernetesRegistry) GetUrl() common.URL {
+	return *r.URL
+}
+
+func (r *kubernetesRegistry) IsAvailable() bool {
+
+	select {
+	case <-r.done:
+		return false
+	default:
+		return true
+	}
+}
+
+func (r *kubernetesRegistry) Destroy() {
+
+	if r.configListener != nil {
+		r.configListener.Close()
+	}
+	r.stop()
+}
+
+func (r *kubernetesRegistry) stop() {
+
+	close(r.done)
```

Comment:
pls use a once.do to defeat using error. maybe it would be invoked more than once.

Reply:
the latest commit has been fixed。